### DISCUSSION TITLE
Stack usage in double buffer ?

### DISCUSSION THREAD

**Discussion by Samuel-DEVULDER** [spotlessmind1975/ugbasic#426] [General]
https://github.com/spotlessmind1975/ugbasic/blob/f94f4fd0817fc7b64c1d23527261304ff744e412/ugbc/src/hw/ef936x/double_buffer.asm#L52-L67

It look to me that here we are using the stack pointer as an address register to some data area. This is okay, but when we use it implicitly in RTS, JSR, PSHS & friends instructions we are actually pushing or pulling data not from the stack but from some random data we've just copied.

Is that the real intended meaning ?

**Comment by spotlessmind1975** [comment]
**hello @Samuel-DEVULDER and thank you for the observation!**

First of all, you are absolutely right! In that code, we copy a memory area directly to the stack area. In a nutshell and on the Olivetti Prodest PC128, I was finding it difficult to combine the current graphics system (which works on `page 0`) with a possible additional system, which works on `page 2` and `page 3`.

The problem was that the executable's current memory layout places the stack (at compile time) starting at address `$A000`  (and down), which is part of `page 3`. At this point I decided to move it to the address `$2000`, as the video memory would be no longer used once the `DOUBLE BUFFER` is activated... but I could not do it at compile time because, until double buffering is activated, that area is used!

I issued this problem on [ugBASIC's Facebook group](https://www.facebook.com/groups/867151224158600).

@dinoflorenzi proposed an assembly code that would dynamically move the stack from location A (`$A000`) to location B (`$1FFF`). The code not only moves the stack pointer with `LDS` instruction but also it copies the current stack, so that execution can continue smoothly. Infact, if you see the previous instructions, you will notice that, as a preliminary step, there is a move of the stack from `$A000` (where it starts) to `$2000 (where it arrives).

Having said all this, I leave room for @dinoflorenzi 's explanation.

**Comment by Neotenien** [comment]
Hello sir, Hello Sam (I'm Neotenien  Bruno aubin)

Why not replace this code

     LDA ,X 
     STA ,S 
     LEAX -1,X 
     LEAS -1,S 
`

By

     LDA ,X- 
     STA ,S- 

You'll gain some precious cycle!

**Comment by Samuel-DEVULDER** [comment]
Because the addressing mode ",x-"  simply does not exists. One can only pre-decrement or post-increment. No post-decrement on the 6809.

Moreover I have not yet looked at optimizing this code because it is only called once. There are probably many other optimizable places to look first.
